internal/terraform/workspace/workspace.go

```diff
@@ -0,0 +1,284 @@
+package workspace
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/terraform-ls/internal/terraform/discovery"
+	tferr "github.com/hashicorp/terraform-ls/internal/terraform/errors"
+	"github.com/hashicorp/terraform-ls/internal/terraform/exec"
+	"github.com/hashicorp/terraform-ls/internal/terraform/lang"
+	"github.com/hashicorp/terraform-ls/internal/terraform/schema"
+)
+
+type workspace struct {
+	ctx            context.Context
```

Comment:
this seems like an issue to attach a context to a struct, you shouldn't need to

Reply:
The reason for the persistence here is to allow watcher to ask for schema refresh without having to pass its own context. The watcher remains global (to help centralising paths we need to watch across the whole hierarchy) so it is designed to outlive a workspace, but operations on the workspace should be cancellable when e.g. workspace cache is removed (not possible today, but I don't see why not implement it later).

Reply:
That said I don't like persisting it either, I just couldn't come up with a better solution.